pkg/typedef/typedef.go

```diff
@@ -197,32 +199,78 @@ const (
 	CacheArrayLen
 )
 
+func splitString(str, delimiter string) func(func(int, string) bool) {
+	lastPos := 0
+	delLen := len(delimiter)
+	return func(yield func(int, string) bool) {
+		for i := 0; ; i++ {
+			pos := strings.Index(str[lastPos:], delimiter)
+
+			if pos == -1 || str[lastPos:] == "" {
+				yield(-1, str[lastPos:])
+
+				break
+			}
+
+			if str[lastPos:lastPos+pos] == "" || !yield(i, str[lastPos:lastPos+pos]) {
+				break
+			}
+
+			lastPos += pos + delLen
+		}
+	}
+}
+
+var builderPool = &sync.Pool{
+	New: func() any {
+		builder := &strings.Builder{}
+
+		builder.Grow(1024)
+
+		return builder
+	},
+}
+
 func prettyCQL(query string, values Values, types Types) string {
 	if len(values) == 0 {
 		return query
 	}
 
-	k := 0
-	out := make([]string, 0, len(values)*2)
-	queryChunks := strings.Split(query, "?")
-	out = append(out, queryChunks[0])
-	qID := 1
-	for _, typ := range types {
-		tupleType, ok := typ.(*TupleType)
-		if !ok {
-			out = append(out, typ.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
-			continue
+	out := builderPool.Get().(*strings.Builder)
+	defer func() {
+		out.Reset()
+		builderPool.Put(out)
+	}()
+
+	next, stop := iter.Pull2(splitString(query, "?"))
+
+	for {
+		i, str, more := next()
+
+		_, _ = out.WriteString(str)
+
+		if !more || i == -1 {
+			stop()
+			break
 		}
-		for _, t := range tupleType.ValueTypes {
-			out = append(out, t.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
+
+		switch ty := types[i].(type) {
+		case *TupleType:
+			for count, t := range ty.ValueTypes {
+				_, _ = out.WriteString(t.CQLPretty(values[count]))
+
+				_, str, more = next()
+				if !more {
+					stop()
+					break
+				}
+
+				_, _ = out.WriteString(str)
+			}
+		default:
+			_, _ = out.WriteString(ty.CQLPretty(values[i]))
 		}
 	}
```

Comment:
It is so confusing, prior variant was way easier to understand.
Also I want to point out that PrettyCQL should throw an error when there is a missmatch between variables and a query.
We should dig into these cases and find out why it is happening.

Reply:
> It is so confusing, prior variant was way easier to understand.
I agree, it's a bit confusing using the new iterators.

> Also I want to point out that PrettyCQL should throw an error when there is a missmatch between variables and a query.
Currently `PrettyCQL` on every type just `panic`s when encounters value that is not expected, if I change the interface to return also an `error`, the whole chain has to be changed. What should be the behavior when a mismatch occurs? Just end the problem, or continue with it, and report it?

Reply:
I would prefer it to fail.

Reply:
@dkropachev Should it be a separate pull request or will be fine if I do it on this one?

Reply:
Lets then merge this one first, after my benchmark is finished and then I'll send another one as it will be a big change, `PrettyCQL` is used in a lot of places.
Best approach I think, is to return an error from it and not crash, as that could be easier for handling down the line.

Reply:
But then changes in `prettyCQL` does not make sense because you will need to redo it again.

Reply:
So you suggest to panic?

Reply:
No, what I meant is that better to keep all the changes to `prettyCQL` in one PR